mirage/stub/dns_stub_mirage.ml

```diff
@@ -0,0 +1,401 @@
+         - unless q in transit (this to-be-done if it is worth it (is it?))
+         - a fresh hdr, q is generated and sent to the recursive resolver
+         - now hdr, q is registered to be awaited for
+         -- we can either signal the request task once we found something,
+            or preserve the original hdr, q together with ip and port
+     - a reply goes out hdr, q, answer
+
+     the "Client" is only concerned about the connection to the resolver, with
+     multiplexing.
+
+     the current API is:
+      dns_client calls connect .. -> flow
+                       send flow data
+                       recv flow (* potentially multiple times *)
+
+     i.e. our flow being (int * _):
+       connect <nothing>
+       send (id, _) data <- id <- data[2..3]
+       recv (id, _) <- registers condition in N[id] ; waits ; removes condition
+
+     or phrased differently:
+       a recv_loop reads continously, whenever a full packet is received,
+        N[id] is woken up with the packet
+  *)
+
+  let inc =
+    let f = function
+      | `Udp_queries -> "udp-queries"
+      | `Tcp_queries -> "tcp-queries"
+      | `Queries -> "queries"
+      | `Decoding_errors -> "decoding-errors"
+      | `Tcp_connections -> "tcp-connections"
+      | `Recursive_queries -> "recursive-questions"
+      | `Recursive_answers -> "recursive-answers"
+      | `Recursive_connections -> "recursive-connections"
+      | `Authoritative_answers -> "authoritative-answers"
+      | `Authoritative_errors -> "authoritative-errors"
+      | `Reserved_answers -> "reserved-answers"
+      | `On_update -> "on-update"
+      | `Resolver_queries -> "resolver-queries"
+      | `Resolver_answers -> "resolver-answers"
+      | `Resolver_nodata -> "resolver-nodata"
+      | `Resolver_nodomain -> "resolver-nodomain"
+      | `Resolver_servfail -> "resolver-servfail"
+      | `Resolver_notimp -> "resolver-notimplemented"
+    in
+    let metrics = Dns.counter_metrics ~f "stub-resolver" in
+    (fun x -> Metrics.add metrics (fun x -> x) (fun d -> d x))
+
+  module IM = Map.Make(struct
+      type t = int
+      let compare : int -> int -> int = fun a b -> compare a b
+    end)
+
+  module Client = struct
+    module Transport : Dns_client.S
+      with type stack = S.t
+       and type +'a io = 'a Lwt.t
+       and type io_addr = Ipaddr.V4.t * int = struct
+      type stack = S.t
+      type io_addr = Ipaddr.V4.t * int
+      type ns_addr = [`TCP | `UDP] * io_addr
+      type +'a io = 'a Lwt.t
+
+      type t = {
+        rng : (int -> Cstruct.t) ;
+        nameserver : ns_addr ;
+        stack : stack ;
+        mutable flow : S.TCPV4.flow option ;
+        mutable requests : (Cstruct.t, [ `Msg of string ]) result Lwt_condition.t IM.t ;
+      }
+      type flow = { t : t ; mutable id : int }
+
+      let create
+          ?rng
+          ?(nameserver = `TCP, (Ipaddr.V4.of_string_exn "91.239.100.100", 53))
```

Comment:
that sounds like an excellent idea we should investigate further to avoid the duplications (maybe open a separate issue, since this is not in scope for this PR)?

Reply:
Opened #210